### PR TITLE
rpc: Require solvability in importmulti if importing more than the scriptPubKey

### PR DESCRIPTION
**Pull request by achow101**
Currently, in `importmulti`, it is possible to import not enough information to make a scriptPubKey solvable. It is also possible to import more information than necessary to make a scriptPubKey solvable. This PR changes this to require that an import must be either just the scriptPubKey or include everything necessary to make it solvable without any extra data.

This is built on top of #14454.

### CODE REVIEW

**Comment by sipa** [comment]
Concept ACK

File: src/wallet/rpcdump.cpp [hunk 1]
```diff
@@ -805,6 +805,111 @@ UniValue dumpwallet(const JSONRPCRequest& request)
     return reply;
 }
 
+struct UsageMaps {
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: Any reason to not move these to `AllUsedKeyStore`?

Comment (promag) [reply src/wallet/rpcdump.cpp]: Got it, thanks.

File: src/wallet/rpcdump.cpp [hunk 2]
```diff
@@ -805,35 +805,190 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
+    WatchKeyUseMap map_watch_key_use;
+    ScriptUseMap map_script_use;
+    KeyUseMap map_key_use;
+};
+
+// Class which remembers whether every script and public key was used
+class AllUsedKeyStore : public CBasicKeyStore
+{
+private:
+    struct UsageMaps* usage = new UsageMaps();
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: This is leaking.

Comment (promag) [reply src/wallet/rpcdump.cpp]: Nit, could remove struct?

File: src/wallet/rpcdump.cpp [hunk 3]
```diff
@@ -855,217 +1010,187 @@ static UniValue ProcessImport(CWallet * const pwallet, const UniValue& data, con
...
         }
 
-        // Invalid P2SH redeemScript
-        if (isP2SH && !IsHex(strRedeemScript)) {
-            throw JSONRPCError(RPC_INVALID_ADDRESS_OR_KEY, "Invalid redeem script");
-        }
-
-        // Process. //
+        CScript scriptpubkey_script = script;
+        CTxDestination scriptpubkey_dest = dest;
```
Comment (practicalswift) [inline src/wallet/rpcdump.cpp]: `scriptpubkey_dest` is unused?

File: src/wallet/rpcdump.cpp [hunk 4]
```diff
@@ -805,35 +805,190 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
+};
+
+// Class which remembers whether every script and public key was used
+class AllUsedKeyStore : public CBasicKeyStore
+{
+private:
+    struct UsageMaps* usage = new UsageMaps();
+public:
+    bool AllUsed() const;
+    bool GetPubKey(const CKeyID &address, CPubKey& vchPubKeyOut) const override;
```
Comment (practicalswift) [inline src/wallet/rpcdump.cpp]: Make the parameters names match between declaration and definition (`CPubKey& vchPubKeyOut` vs `CPubKey &pubkey_out`) :-)

File: src/wallet/rpcdump.cpp [hunk 5]
```diff
@@ -805,35 +805,190 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
+    bool AddCScript(const CScript& redeemScript) override;
+    bool GetCScript(const CScriptID &hash, CScript& redeemScriptOut) const override;
```
Comment (practicalswift) [inline src/wallet/rpcdump.cpp]: Same here: `CScript& redeemScriptOut` vs `CScript& script_out` :-)

File: src/wallet/rpcdump.cpp [hunk 6]
```diff
@@ -805,35 +805,190 @@ UniValue dumpwallet(const JSONRPCRequest& request)
...
+    bool AddWatchOnly(const CScript &dest) override;
+    bool AddKeyPubKey(const CKey& key, const CPubKey &pubkey) override;
+    bool GetKey(const CKeyID &address, CKey &keyOut) const override;
+};
+
+bool AllUsedKeyStore::GetPubKey(const CKeyID &address, CPubKey &pubkey_out) const
+{
+    LOCK(cs_KeyStore);
+    if (CBasicKeyStore::GetPubKey(address, pubkey_out)) {
+        usage->map_watch_key_use[address] = true;
```
Comment (promag) [inline src/wallet/rpcdump.cpp]: You could invert the logic and keep track of the unused
 - when adding a key, add to the unused set
 - when reading a key, remove from the unused set

Then `AllUsed` could look like (s/usage/unused):
```cpp
    return 
        unused->watch_keys.empty() &&
        unused->scripts.empty() &&
        unused->keys.empty();
```

I could be missing some detail.

**Pull request closed without merging**